### ISSUE TITLE
Push notifications not working if register on app launch

### ISSUE THREAD

**Issue by vpilkevych** [firebase/firebase-ios-sdk#9003]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 13.1.0
  * Firebase SDK version: 8.8.0, 8.9.0
  * Installation method: `CocoaPods`
  * Firebase Component: Messaging
  * Target platform(s): `iOS`

### [REQUIRED] Step 2: Describe the problem

The app doesn't receive push notifications sent via Firebase Cloud Messaging if certain conditions are met:
- The app is compiled with a distribution provisioning profile, e.g Ad-Hoc, so that the app uses APNs production environment. It works fine when built with a debug provisioning profile (and thus with sandbox APNs environment).
- The app configures Firebase, sets the messaging delegate and registers for APNs on its launch. When this happens Firebase outputs the following warning message which must be a result of its internal call because the app  doesn't request the FCM token at this point:
> 8.9.1 - [Firebase/Messaging][I-FCM002022] APNS device token not set before retrieving FCM Token for Sender ID '532814239450'. Notifications to this FCM Token will not be delivered over APNS.Be sure to re-retrieve the FCM token once the APNS device token is set.
- The app is started for the first time after the installation (e.g. via Xcode or from App Distribution).
- It may be required that previously the app with the same Firebase version was installed, registered with APNs and deleted. It seems it may something to do with the state persisted in the Keychain.

#### Steps to reproduce:

1. Install the app built with Ad-Hoc provisioning profile on the device via Xcode.
2. Run the app for the first time and let the app to configure Firebase and register for APNs on its launch.
3. Provide the APNs device token to Firebase messaging.
4. See the mentioned warning in Xcode output pane.
5. When Firebase messaging provides the FCM token send it to the backend.
6. Send a push notification from the backend to Firebase Cloud Messaging using the FCM token.
7. The notification is not delivered.

Sending the notification via APNs directly at this point works fine. Firebase is configured with the APNs Authentication Key and it works fine with the very same project and same conditions if Firebase Messaging v7.0.0 is used.

#### Relevant Code:

AppDelegate.swift:
```
import FirebaseMessaging

...

func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    FirebaseApp.configure()
    Messaging.messaging().delegate = self
    application.registerForRemoteNotifications()
}

func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    Messaging.messaging().apnsToken = deviceToken
}
```

**Comment by rizafran** [comment]
Thanks for reporting this, @vpilkevych. I see that you're unable to receive notifications upon installing the app. Could you confirm if the issue doesn't happen if the app is already installed (or opening the app after quitting it)? It looks like the use case and warning message you shared is related to this [Unity issue](https://github.com/firebase/quickstart-unity/issues/1176).

**Comment by vpilkevych** [comment]
@rizafran yes I can confirm that the issue doesn't happen if the app is already installed or if it is restarted by force-quitting it. It does indeed resemble the mentioned unity issue, although i haven't noticed the warning message being mentioned there.

**Comment by rizafran** [comment]
Thanks for confirming, @vpilkevych. It looks like we're having the same behavior using the quickstart. With this, I'll notify an engineer and see what we can do here.

**Comment by chliangGoogle** [comment]
@leojaygoogle This seems related to the issue we discussed, would you take a lead and investigate?

**Comment by bluesource** [comment]
Still happening with  8.10.0

**Comment by omar-saadoun** [comment]
This is still happening with 8.10.0

**Comment by mahgo** [comment]
Any update on this?

**Comment by rizafran** [comment]
Hi, the issue is being investigated. We're tracking this internally at b/209693808, and will follow up if we need more information. Thanks.

**Comment by jerry-inc** [comment]
I am experiencing the same thing. 

When I run the app from Xcode, I can receive notifications normally. However, When I make a build and try to test it from TestFlight, I do not receive any notification.

**Comment by Asadeck** [comment]
I am not even receiving the notifications even in debug mode

**Comment by chliangGoogle** [comment]
We've fixed the issue at https://github.com/firebase/firebase-ios-sdk/pull/9152. Please wait for next Firebase release and test to let me know if the issue still happens.

**Comment by paulb777** [comment]
Marking as fixed based on the #9152 fix that is being staged for 8.11.0 due out next week.
